### PR TITLE
Only testing doc when only doc is modified

### LINKED ISSUE #47: skip tests when only docs/ have been modified
Often PRs only modify documentation (the top `doc/` directory) so there isn't a point in running the unit or integration tests, we should create a script that allows us to determine if a PR is exclusively modifying `doc/` and nothing else.

The API should work as follows

```sh
python -m pkgmt.ensure_modified --base-branch {BASE} --only {PATH}
```

For example, to check if the current branch only modified `doc/` with respect to the `main` branch, I can run:

```sh
python -m pkgmt.ensure_modified --base-branch main --only doc/
```

If the current branch has only modified `doc/` the command should finish with exit code 0, if files outside `doc/` have been modified, it should finish with exit code 1.

Inside Python, we can use `subprocess` to run `git` commands and find out. We already have a similar workflow that checks if [CHANGELOG.md has been modified](https://github.com/ploomber/contributing/blob/main/sample-github-workflows/changelog.yml) but the logic is written in bash, for this issue, we should write the logic in Python as it's easier to maintain. Sample git commands to determine the modified files can be found [here](https://github.com/ploomber/contributing/blob/d2004250dea96e06ab120db0e50cb315b19c3b78/sample-github-workflows/workflow-edits.yml#L35).

Note that we need to make a slight modification and add the `...` operator, since we're only interested in changes introduced by the current branch (not the ones introduced by the `BASE_BRANCH`:

```sh
git diff --exit-code "$BASE_BRANCH..." doc/
```

### PR DESCRIPTION
**Pull request by mehtamohit013**
Closes #47 

TO DO:

- [X] implement python -m pkgmt ensure_modified
- [X] add unit tests
- [X] implement sample github workflow (you can test in the [dummy](https://github.com/ploomber/dummy) repository, ask @edublancas for edit access)
- [ ] open PR with the finished workflow to this repo https://github.com/ploomber/contributing

### CODE REVIEW

**Comment by mehtamohit013** [comment]
Hey @edublancas ,
please take a look at the `modified.py` and check if it satisfies all the requirements.

I have also added some test that works fine locally, but not with GitHub actions. Seems like I have to use tmp_packages in `conftest.py`. Any suggestion?

Sample commands:
Context: the current branch has 3 diff: folder `doc`, `another_doc` and file `a.txt`
1. ` pkgmt ensure-modified -b main  -ex doc -ex another_doc -ex a.txt  --debug`
    Output: `0`
    
2. ` pkgmt ensure-modified -b main  -ex another_doc -ex a.txt  --debug`
    Output: `1`

File: src/pkgmt/cli.py [hunk 1]
```diff
@@ -174,3 +175,29 @@ def lint():
 
     if returncode:
         raise SystemExit("Error linting")
+
+
+@cli.command()
+@click.option(
```
Comment (edublancas) [inline src/pkgmt/cli.py:181]: please move this under pkgmt/modified.py, so we can execute it with `python -m pkgmt.modified` (we don't want to add this to the main CLI since this is mostly for CI purposes)

File: src/pkgmt/cli.py [hunk 2]
```diff
@@ -174,3 +175,29 @@ def lint():
...
+    "--base-branch",
+    default="main",
+    help="Base branch to compare against",
+)
+@click.option(
+    "-ex",
+    "--exclude-path",
+    default=["doc"],
+    multiple=True,
+    help="Path to exclude from git diff. Can be used multiple times eg: -p p1 -p p2",
```
Comment (edublancas) [inline src/pkgmt/cli.py:192]: looks like this will evaluate if I changed *anything* except what's listed. but that'd be pretty cumbersome if we want to check whether I *exclusively* changed `doc/` because it'd require me to list all directory. say I have directories a/,..z/ and doc/, if I only want to check if I changed doc/ I need to run:

```
pkgmt ensure-modified --base-branch master -p a -p b -p c ... -p z
```

what we want is to check whether `doc/` is the only one that has been modified by only listing that one:

```
pkgmt ensure-modified --base-branch master --exclusive doc
```

this should fail (exit code 1) if I modified anything outside doc/. and exit with code 0 if I only modified things inside doc/

and this:

```
pkgmt ensure-modified --base-branch master --exclusive doc --exclusive another
```

this would fail if I modified anything outside doc/ or /another

File: src/pkgmt/cli.py [hunk 3]
```diff
@@ -174,3 +175,29 @@ def lint():
...
+
+@cli.command()
+@click.option(
+    "-b",
+    "--base-branch",
+    default="main",
+    help="Base branch to compare against",
+)
+@click.option(
+    "-ex",
```
Comment (edublancas) [inline src/pkgmt/cli.py:188]: the options are inconsistent, here it says `-ex` but the example says `-p`

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/pkgmt/cli.py (hunk 2 above)
Comment (mehtamohit013) [reply src/pkgmt/cli.py:192]: True, that is one way to see it.

Another way would be:
Suppose we want to check whether `doc` folder is modified in the codebase, so if I run `pkgmt ensure-modified -b main  -ex doc` and only `doc` folder is modified, excluding it from `git diff` and compare it with `main` branch would result in the output of `0` as **nothing is modified**. But as we know, there is a commit, **so something must be modified(underlying assumption)**, as **excluding doc**, results in no modification in the codebase, therefore **doc must have been modified**

Output: `0` : "**Nothing is modified excluding ...**" is equivalent to "**something is modified in excluded ...**" (Run only doc tests)
`1` : "**Something is modified excluding ...**" is **not** equivalent to "**nothing is modified in excluded ...**" (Run all tests)

@edublancas

**Comment by mehtamohit013** [comment]
@edublancas 

> open PR with the finished workflow to this repo https://github.com/ploomber/contributing

Also, can you please elaborate more on what you want to add to contributing guidelines?

File: src/pkgmt/modified.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                f"Path has been modified with respect to '{base_branch}'\n"
+                f"Excluding paths: {exclude_path}"
+            )
+            print(f"Return code: {err.returncode}")
+            print(f"Output: {err.output}")
+        return 1
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
```
Comment (edublancas) [inline src/pkgmt/modified.py:29]: you can still use click and put your code here, but since you already migrated to ArgumentParser, let's keep it as it is

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/pkgmt/cli.py (hunk 2 above)
Comment (edublancas) [reply src/pkgmt/cli.py]: I'm not following this logic. Let's implement it, test it in dummy and see how it works

File: src/pkgmt/modified.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        default=["doc"],
+        nargs="+",
+        help="Path to exclude from git diff."
+        "Can be used multiple times eg: -ex p1 -ex p2",
+    )
+    parser.add_argument("--debug", action="store_true", help="Print debug info")
+
+    args = parser.parse_args()
+    returncode = check_modified(args.base_branch, args.exclude_path, debug=args.debug)
+    print(returncode)
```
[the remark this answers is not in the document]
Comment (mehtamohit013) [reply src/pkgmt/modified.py]: When using `sys.exit`, in the case of exit-code `1`, the workflow was terminated and no further command was executed. However, what I initially have in mind is to capture the `stdout`, and use the `if...else` condition statement in the bash script to identify whether to run full tests or only doc tests.

So, for now, I am using print, to test out the logic. We can come to this later

**Comment by mehtamohit013** [comment]
@edublancas 

I have written a template based on the above logic [here](https://github.com/ploomber/dummy/blob/pkgmt_modi/.github/workflows/doc_edited.yml)

Testing: 
For testing the logic, I made `pkgmt_modi` branch with a new GitHub workflow. After that, I made three branches with the following commits on top of `pkgmt_modi` branch:

- `pkgmt_modi1`: 
    File Changed: `Modified doc/a.txt`
    Output: [GitHub Actions]( https://github.com/ploomber/dummy/actions/runs/5135591724/jobs/9241232294#step:6:171)
    
- `pkgmt_modi2`: 
    File Changed: `Modified a.txt`
    Output: [GitHub Actions](https://github.com/ploomber/dummy/actions/runs/5135626218/jobs/9241316615#step:6:171)
    
- `pkgmt_modi3`: 
    File Changed: `Modified a.txt, doc/a.txt`
    Output: [GitHub Actions](https://github.com/ploomber/dummy/actions/runs/5135657928/jobs/9241387207#step:6:171)    
    
TLDR: Tests are working as intended

**Review by edublancas — CHANGES_REQUESTED** [review]
Update: nevermind. I realized the branch was hardcoded

am I testing this wrong?

I copied your workflow: https://github.com/ploomber/dummy/blob/main/.github/workflows/doc.yml

added a new file under doc/ https://github.com/ploomber/dummy/commit/32a4fa3c549b316333a4a2fb2a3173634e09648e

but the output still says: executing all tests

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/pkgmt/modified.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import subprocess
+import argparse
+
+
+def check_modified(base_branch, exclude_path, debug=False):
+    # https://stackoverflow.com/questions/4380945
+    cmd = f"git diff --exit-code {base_branch}... -- ."
+
```
Comment (edublancas) [inline src/pkgmt/modified.py:8]: also, I thin let's rename this to `fail_if_modified` to imply that if anything has been modified (excluding the --exclude-path), this will fail, aka return exit code 1

**Review by edublancas — CHANGES_REQUESTED** [review]

File: tests/conftest.py [hunk 7]
```diff
@@ -57,3 +57,31 @@ def tmp_another_package(root, tmp_empty):
...
+    subprocess.run(["git", "init"])
+    subprocess.check_call(["git", "config", "commit.gpgsign", "false"])
+    subprocess.check_call(["git", "config", "user.email", "ci@ploomberio"])
+    subprocess.check_call(["git", "config", "user.name", "Ploomber"])
+    subprocess.run(["git", "checkout", "-b", "main"])
+    subprocess.run(["git", "add", "--all"])
+    subprocess.run(["git", "commit", "-m", "init-commit-message"])
+
+    subprocess.run(["git", "checkout", "-b", "test_modified_doc"])
+    subprocess.run(["mkdir", "-p", "test_doc1"])
```
[the remark this answers is not in the document]
Comment (mehtamohit013) [reply tests/conftest.py]: I have converted all the calls to `pathlib.Path`, but Windows tests are still failing. For more info, check the tests [here](https://github.com/ploomber/pkgmt/actions/runs/5147192131/jobs/9267273488?pr=49#step:6:28)

[the remark this answers is not in the document]
Comment (edublancas) [reply tests/conftest.py]: yeah, some stuff isn't compatible with windows, so no need to add windows and macos to the testing configuration. I suggested changing it to pathlib because at some point we might want to support those OS as well

[the remark this answers is not in the document]
Comment (mehtamohit013) [reply tests/conftest.py]: So, I commented all those tests, but `test_check_modified` is failing. It seems to me more of a `git` issue because the `debug` output says `modified.txt` has been deleted and a new file `modified.txt` has been created. (Output [here](https://github.com/ploomber/pkgmt/actions/runs/5147192131/jobs/9267273488?pr=49#step:6:77))

So, is it ok for now to skip this part, the part of debugging windows test, I have already added converted the calls to `path.Pathlib`?

File: src/pkgmt/modified.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import subprocess
```
[the remark this answers is not in the document]
Comment (mehtamohit013) [reply src/pkgmt/modified.py]: [Template](https://github.com/ploomber/dummy/blob/pkgmt_modi/.github/workflows/doc_edited.yml)

One can use `if: steps.check_doc.outcome == 'failure'` to determine whether only doc has been modified or not

**Comment by mehtamohit013** [comment]
@edublancas 
[Template](https://github.com/ploomber/dummy/blob/pkgmt_modi/.github/workflows/doc_edited.yml)

With `sys.exit`
Testing: 

- `pkgmt_modi1`: 
    File Changed: `Modified doc/a.txt`
    Branch Compared: `origin/pkgmt_modi`
    Output: [GitHub Actions](https://github.com/ploomber/dummy/actions/runs/5157895090/jobs/9290896742)
    
- `pkgmt_modi2`: 
    File Changed: `Modified a.txt`
    Branch Compared: `origin/pkgmt_modi`
    Output: [GitHub Actions](https://github.com/ploomber/dummy/actions/runs/5157925148/jobs/9290963706)
    
- `pkgmt_modi3`: 
    File Changed: `Modified a.txt, doc/a.txt`
    Branch Compared: `origin/pkgmt_modi`
    Output: [GitHub Actions](https://github.com/ploomber/dummy/actions/runs/5157937831/jobs/9290989715)    
    
TLDR: Tests are working as intended

File: tests/conftest.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mehtamohit013) [reply tests/conftest.py]: I think it maybe because of not using `' '` in exclude path like `':^doc'` instead of `:^doc` (due to this [error](https://github.com/ploomber/pkgmt/actions/runs/5157612645/jobs/9290253296#step:6:231)), I will try this approach tomorrow

[the remark this answers is not in the document]
Comment (mehtamohit013) [reply tests/conftest.py]: Nope not working
[Github Actions](https://github.com/ploomber/pkgmt/actions/runs/5157612645)

Some links: 
[Link1](https://github.com/ploomber/pkgmt/actions/runs/5157612645/jobs/9290253296#step:6:214)
[Link2](https://github.com/ploomber/pkgmt/actions/runs/5157612645/jobs/9290253296#step:6:231)

**Comment by mehtamohit013** [comment]
@edublancas 

[Template](https://github.com/ploomber/dummy/blob/pkgmt_modi/.github/workflows/doc_edited.yml)

With windows testing and `sys.exit`
Testing: 

- `pkgmt_modi1`: 
    File Changed: `Modified doc/a.txt`
    Branch Compared: `origin/pkgmt_modi`
    Output: [GitHub Actions](https://github.com/ploomber/dummy/actions/runs/5177602495)
    
- `pkgmt_modi2`: 
    File Changed: `Modified a.txt`
    Branch Compared: `origin/pkgmt_modi`
    Output: [GitHub Actions](https://github.com/ploomber/dummy/actions/runs/5177612096)
    
- `pkgmt_modi3`: 
    File Changed: `Modified a.txt, doc/a.txt`
    Branch Compared: `origin/pkgmt_modi`
    Output: [GitHub Actions](https://github.com/ploomber/dummy/actions/runs/5177622635)    
    
TLDR: Tests are working as intended on ubuntu and macos. **Not working for windows**

**Comment by edublancas** [comment]
@mehtamohit013: just to confirm what we discussed earlier today. since windows support is giving too much trouble, let's not worry about it for now

**Comment by mehtamohit013** [comment]
> @mehtamohit013: just to confirm what we discussed earlier today. since windows support is giving too much trouble, let's not worry about it for now

@edublancas 
Ok, then, I believe I have resolved all the issues and the PR is ready to be reviewed and merge

**Review by edublancas — APPROVED** [review]
